### ISSUE TITLE
please remove the -pthread option in compile/link script to support PGI compilers

### ISSUE THREAD

**Issue by walkup** [open-mpi/ompi#2331]
The openmpi wrapper scripts (mpicc, mpicxx, mpif90, etc.) include a "-pthread" option, which causes compilation and/or linking failures with PGI compilers (pgcc, pgc++, pgf90).  That option is not needed on current Linux systems, and is fatal for some compilers, including the PGI compilers.  Please remove the "-pthread" compile/link option from the compile/link scripts.

**Comment by ggouaillardet** [comment]
@walkup we previously got a similar report.
what happened it the ```-pthread``` flag was pulled from some external libtool generated files built with gcc
(iirc, slurm was compiled with gcc, and hence Open MPI compiled with PGI compiler pulled the ```-pthread``` flag)
are you running into a similar case ?

workarounds are
- build all your dependencies with PGI
- edit the ```*.la``` files from your dependencies and remove the ```-pthread``` flag
- remote the ```*.la``` files from your dependencies

an other option is to patch libtools so ```-pthread``` flag is ignored by PGI compilers.
that would definitely solve your issue, but might silently generate non thread-safe code so i am not too comfortable with that.

**Comment by walkup** [comment]
Thanks.  I understand that the "-pthread" option got in there for historical reasons.  However, it is an option that is specific to gcc, and if you recursively grep for REENTRANT and THREAD_SAFE in openmpi source, including all header files, you will find that the "-pthread" option is not required when building with gcc, even when building support for MPI_THREAD_MULTIPLE.  One can safely eliminate all references to "-pthread" in the openmpi build infrastructure.  Including -lpthread in the list of libraries that get linked is a good idea ... that is not the same as adding "-pthread" as a linker option, because "-pthread" can cause linking to fail.  In any case, the "-pthread" option should never be passed to the compile/link scripts because those may very well be using a different compiler (via OMPI_CC, OMPI_CXX, OMPI_FC), where "-pthread" may not be a valid option, and can cause compilation and/or linking faliure.  I know how to fix the scripts after doing a build/install ... but it would be preferable for the default openmpi build/install process to just work with any compiler that a user might choose at compile time.

**Comment by ggouaillardet** [comment]
Open MPI needs pthreads, and hence need to tell the compiler to build thread safe code.
if not, ```errno``` might be a unique global variable instead of a per thread variable.
/* we did run into this, not on linux nor with GNU compilers iirc */
that can be achieved with ```-pthread``` with ```gcc```, several other options are available, including ```-D_REENTRANT -lpthread```
bottom line, ```-pthread``` or a similar option __is__ required to correctly build Open MPI.

if you use ```OMPI_CC``` and friends, then you basically hope things will work
/* if you built with ```gfortran``` and have ```OMPI_FC``` point to a non ```GNU``` compiler (or even a different version of gfortran), this is just hopeless */
bottom line, it is simply impossible to "just work with any compiler chosen at compile time"

IMNSHO, ideally, ```libtool``` would automagically translate the ```-pthread``` option to something that is working for the compiler. that would fix the build issue when building with ```PGI``` compilers and using external ```.la``` files (such as SLURM) generated with ```GNU``` compilers.

**Comment by correaa** [comment]
In case it matters, `nvcc` has the same problem. It rejects the `pthread` option and refuses to compile.
Supposedly `nvcc` doesn't need the option but effectively acts if the option is active.

**Comment by gpaulsen** [comment]
@janjust Can we close this?  I believe that the `-pthread` compiler issue has been resolved in new (PGI / nvcc) compiler versions 16 and beyond?  Can you confirm and close if true?

**Comment by janjust** [comment]
@cparrot73 is probably better suited to answer this.

@open-mpi/cuda for vis

**Comment by walkup** [comment]
I can confirm that there is no problem with -pthread when using the current
NVIDIA HPC Toolkit, with nvfortran as the underlying Fortran compiler.  I
think this "issue" can be closed.

Regards,
Bob Walkup ***@***.***, 914-945-1512)



From:	"Geoff Paulsen" ***@***.***>
To:	"open-mpi/ompi" ***@***.***>
Cc:	"walkup" ***@***.***>, "Mention"
            ***@***.***>
Date:	11/04/2021 02:31 PM
Subject:	[EXTERNAL] Re: [open-mpi/ompi] please remove the -pthread
            option in compile/link script to support PGI compilers (#2331)



@janjust Can we close this? I believe that the -pthread compiler issue has
been resolved in new (PGI / nvcc) compiler versions 16 and beyond? Can you
confirm and close if true? — You are receiving this because you were
mentioned. ‍‍‍‍


@janjust Can we close this? I believe that the -pthread compiler issue has
been resolved in new (PGI / nvcc) compiler versions 16 and beyond? Can you
confirm and close if true?


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
